tests_e2e/playwright/run_notebook.py

```diff
@@ -0,0 +1,106 @@
+        self.nav.initialize
+
+    def run_notebook(self, path, expected_output_text, conda_env, runtime=30000):
+        """Run jupyter notebook and check for expected output text anywhere on
+        the page.
+
+        Note: This will look for and exact match of expected_output_text
+        _anywhere_ on the page so be sure that your text is unique.
+
+        Conda environments may still be being built shortly after deployment.
+
+        conda_env: str
+            Name of conda environment. Python conda environments have the
+            structure "conda-env-nebari-git-nebari-git-dashboard-py" where
+            the actual name of the environment is "dashboard".
+        """
+        logger.debug(f">>> Running notebook: {path}")
+        filename = Path(path).name
+
+        # navigate to specific notebook
+        file_locator = self.nav.page.get_by_text("File", exact=True)
+
+        file_locator.wait_for(
+            timeout=self.nav.wait_for_server_spinup,
+            state="attached",
+        )
+        file_locator.click()
+        self.nav.page.get_by_role("menuitem", name="Open from Path…").get_by_text(
+            "Open from Path…"
+        ).click()
+        self.nav.page.get_by_placeholder("/path/relative/to/jlab/root").fill(path)
+        self.nav.page.get_by_role("button", name="Open", exact=True).click()
+        # give the page a second to open, otherwise the options in the kernel
+        # menu will be disabled.
+        self.nav.page.wait_for_load_state("networkidle")
+        if self.nav.page.get_by_text(
+            "Could not find path:",
+            exact=False,
+        ).is_visible():
+            logger.debug("Path to notebook is invalid")
+            raise RuntimeError("Path to notebook is invalid")
+        # make sure the focus is on the dashboard tab we want to run
+        self.nav.page.get_by_role("tab", name=filename).get_by_text(filename).click()
+        self.nav.set_environment(kernel=conda_env)
+
+        # make sure that this notebook is one currently selected
+        self.nav.page.get_by_role("tab", name=filename).get_by_text(filename).click()
+
+        # restart run all cells
+        self.nav.page.get_by_text("Kernel", exact=True).click()
+        self.nav.page.get_by_role(
+            "menuitem", name="Restart Kernel and Run All Cells…"
+        ).get_by_text("Restart Kernel and Run All Cells…").click()
+
+        # Restart dialog appears most, but not all of the time (e.g. set
+        # No Kernel, then Restart Run All)
+        restart_dialog_button = self.nav.page.get_by_role(
+            "button", name="Restart", exact=True
+        )
+        if restart_dialog_button.is_visible():
+            restart_dialog_button.click()
+
+        output_locator = self.nav.page.get_by_text(expected_output_text, exact=True)
+        output_locator.wait_for(timeout=runtime)  # wait for notebook to run
+        assert output_locator.is_visible()
+
+
+if __name__ == "__main__":
```

Comment:
I don't really understand the purpose of this. The `Navigator` main is a smoke test for `nebari` in general. But this main here runs a fixed notebook similar to what the test is doing right? Since we are also cloning a repo here, we can't really use a local notebook that I want to test. Meaning, this main seems somewhat obsolete. Can we remove or did I miss something?

Reply:
Could you walk me to the process on how to run a local notebook then? IIUC, I would need to

1. Fix the `nebari_url` parameter
2. Set the `USERNAME` / `PASSWORD` environment variables either manually or through `dotenv`
3. Comment out the whole `clone_repo` step
4. Fix the path in `run_notebook` to the local notebook that I want to test.

If I would do the same through a regular test, the steps would be

1. Copy a test like `test_notebook`
2. Comment out the whole `clone_repo` step
3. Fix the path in `run_notebook` to the local notebook that I want to test.
4. Set the `USERNAME` / `PASSWORD` / ... environment variables either manually or through `dotenv`
5. Run `pytest`

Unless I'm missing something here, this is somewhat equivalent. The upside of going the `pytest` route is that you already have written your test, which is what you likely want to do when running a notebook like this.

Reply:
>Comment out the whole clone_repo step

This runs on the nebari instance, which doesn't have the local copy of the repo so this is still needed. 

Other than that, yes, they are equivalent. `pytest` will replicate CI, obviously, but I like to run as main to allow me to use the VS Code debugger. Developer convenience.

Reply:
As discussed offline, I was mistakenly under the impression that both scenarios I mentioned above give the same DevX. I did no know that the VS Code debugger is a mess when debugging tests. Thus, this main has a good reason to exist.